File: 322-coin-change/322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        q = [amount]
        seen = {amount}
        ans = 0
        while q:
            nq = []
            for r in q:
                if r == 0:
                    return ans
                
                if r > 0:
                    for coin in coins:
                        if r - coin in seen:
                            continue
                        nq.append(r - coin)
                        seen.add(r - coin)
            q = nq
            ans += 1
        return -1

File: 322-coin-change/322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int ans = helper(
            coins, amount, new int[amount + 1]
        );
        
        return (ans == Integer.MAX_VALUE) ? -1 : ans;
    }
    
    private int helper(int[] coins, int remaining, int[] dp) {
        if (remaining < 0) 
            return Integer.MAX_VALUE;
        
        if (remaining == 0)
            return 0;
        
        if (dp[remaining] == 0){
            int ans = Integer.MAX_VALUE;
            for (int coin : coins)
                ans = Math.min(ans, helper(coins, remaining - coin, dp));
            dp[remaining] = (ans == Integer.MAX_VALUE) ? Integer.MAX_VALUE : ans + 1;
        }
        
        return dp[remaining];
    }
}